examples/vision/retinanet.py

```diff
@@ -0,0 +1,700 @@
+            reduction="none", name="RetinaNetClassificationLoss"
+        )
+        self._alpha = alpha
+        self._gamma = gamma
+
+    def call(self, y_true, y_pred):
+        cross_entropy = tf.nn.sigmoid_cross_entropy_with_logits(
+            labels=y_true, logits=y_pred
+        )
+        probs = tf.nn.sigmoid(y_pred)
+        alpha = tf.where(tf.equal(y_true, 1.0), self._alpha, (1.0 - self._alpha))
+        pt = tf.where(tf.equal(y_true, 1.0), probs, 1 - probs)
+        loss = alpha * tf.pow(1.0 - pt, self._gamma) * cross_entropy
+        return tf.reduce_sum(loss, axis=-1)
+
+
+class RetinaNetLoss(tf.losses.Loss):
+    """Wrapper to combine both the losses"""
+
+    def __init__(self, num_classes=80, alpha=0.25, gamma=2.0, delta=1.0):
+        super(RetinaNetLoss, self).__init__(reduction="auto", name="RetinaNetLoss")
+        self._cls_loss = RetinaNetClassificationLoss(alpha, gamma)
+        self._box_loss = RetinaNetBoxLoss(delta)
+        self._num_classes = num_classes
+
+    def call(self, y_true, y_pred):
+        y_pred = tf.cast(y_pred, dtype=tf.float32)
+        box_labels = y_true[:, :, :4]
+        box_predictions = y_pred[:, :, :4]
+        cls_labels = tf.one_hot(
+            tf.cast(y_true[:, :, 4], dtype=tf.int32),
+            depth=self._num_classes,
+            dtype=tf.float32,
+        )
+        cls_predictions = y_pred[:, :, 4:]
+        positive_mask = tf.cast(tf.greater(y_true[:, :, 4], -1.0), dtype=tf.float32)
+        ignore_mask = tf.cast(tf.equal(y_true[:, :, 4], -2.0), dtype=tf.float32)
+        cls_loss = self._cls_loss(cls_labels, cls_predictions)
+        box_loss = self._box_loss(box_labels, box_predictions)
+        cls_loss = tf.where(tf.equal(ignore_mask, 1.0), 0.0, cls_loss)
+        box_loss = tf.where(tf.equal(positive_mask, 1.0), box_loss, 0.0)
+        normalizer = tf.reduce_sum(positive_mask, axis=-1)
+        cls_loss = tf.math.divide_no_nan(tf.reduce_sum(cls_loss, axis=-1), normalizer)
+        box_loss = tf.math.divide_no_nan(tf.reduce_sum(box_loss, axis=-1), normalizer)
+        loss = cls_loss + box_loss
+        return loss
+
+
+def add_l2_regularization(weight, alpha=0.0001):
+    """Utility function to add l2 regularization"""
+
+    def _add_l2_regularization():
+        l2_reg = tf.keras.regularizers.l2(alpha)
+        return l2_reg(weight)
+
+    return _add_l2_regularization
+
+
+"""
+## Setting up training parameters
+"""
+
+model_dir = "retinanet/"
+label_encoder = LabelEncoder()
+
+num_classes = 80
+batch_size = 2
+
+learning_rates = [2.5e-06, 0.000625, 0.00125, 0.0025, 0.00025, 2.5e-05]
+learning_rate_boundaries = [125, 250, 500] + [240000, 360000]
+learning_rate_fn = tf.optimizers.schedules.PiecewiseConstantDecay(
+    boundaries=learning_rate_boundaries, values=learning_rates
+)
+
+"""
+## Initializing and compiling model
+"""
+
+loss_fn = RetinaNetLoss(num_classes)
+model = RetinaNet(num_classes)
+
+
+#  Add L2 regularization to all Conv2d weights
```

Comment:
Prefer doing so by passing a `kernel_regularizer` to Conv2D layers at model construction. The setup below is very non-standard.

Reply:
Yes I can do that for the new layers that I am adding, but how do I set the `kernel_regularizer` attribute for the the layers in ResNet50 which is already built?

Reply:
Does it really affect performance negatively? In general L2 regularization on conv kernels is not an effective regularization technique (dropout is preferred)

Reply:
I have not tried training this model without L2 regularization, but I have seen that it does affect the performance for other  detection models, what's the proper way if we do have to go with this?

Reply:
you should use weight decay, not l2 reg (or at least that's how the original paper is done).
Specifically you can rely on tensorflow_addons.optimizers.SGDW

Reply:
@tanzhenyu Yes, I can try retraining the model with `tensorflow_addons.optimizers.SGDW`. 
@fchollet would this be okay because `tensorflow_addons` would be an additional dependency.

Reply:
@tanzhenyu I tried training with `tensorflow_addons.optimizers.SGDW` from scratch, but the model failed to converge. 
IIUC from the official [tutorial](https://www.tensorflow.org/tutorials/keras/overfit_and_underfit) in keras we use l2 regularization to add weight decay.
```
L2 regularization, where the cost added is proportional to the square of the value of the weights 
coefficients (i.e. to what is called the squared "L2 norm" of the weights). L2 regularization is also 
called weight decay in the context of neural networks. 
Don't let the different name confuse you: weight decay is mathematically the exact same as L2 regularization.
```
@fchollet What do we here?